ateam_kenobi/src/skills/line_kick.cpp

```diff
@@ -0,0 +1,120 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#include "line_kick.hpp"
+#include <angles/angles.h>
+#include <vector>
+#include <ateam_geometry/normalize.hpp>
+
+namespace ateam_kenobi::skills
+{
+
+LineKick::LineKick(visualization::OverlayPublisher & overlay_publisher)
+: overlay_publisher_(overlay_publisher),
+  easy_move_to_(overlay_publisher_)
+{
+}
+
+ateam_geometry::Point LineKick::getAssignmentPoint(const World & world)
+{
+  return getPreKickPosition(world);
+}
+
+ateam_msgs::msg::RobotMotionCommand LineKick::runFrame(const World & world, const Robot & robot)
+{
+  const auto pre_kick_position = getPreKickPosition(world);
+
+  overlay_publisher_.drawLine("LineKick_line", {pre_kick_position, target_point_}, "#FFFF007F");
+
+  const auto distance_to_pre_kick = ateam_geometry::norm(robot.pos, pre_kick_position);
+  if (distance_to_pre_kick > 0.05) {
+    if (prev_state_ != State::MoveToPreKick) {
+      easy_move_to_.reset();
+      prev_state_ = State::MoveToPreKick;
+    }
+    return moveToPreKick(world, robot);
+  }
+
+  const auto robot_to_ball = target_point_ - robot.pos;
+  const auto robot_to_ball_angle = std::atan2(robot_to_ball.y(), robot_to_ball.x());
+  if (angles::shortest_angular_distance(robot.theta, robot_to_ball_angle) > 0.02) {
+    if (prev_state_ != State::FaceBall) {
+      easy_move_to_.reset();
+      prev_state_ = State::FaceBall;
+    }
+    return faceBall(world, robot);
+  }
+
+  if (prev_state_ != State::KickBall) {
+    easy_move_to_.reset();
+    prev_state_ = State::KickBall;
+  }
+  return kickBall(world, robot);
+}
+
+ateam_geometry::Point LineKick::getPreKickPosition(const World & world)
+{
+  return world.ball.pos + (kPreKickOffset * ateam_geometry::normalize(
+           world.ball.pos - target_point_));
+}
+
+ateam_msgs::msg::RobotMotionCommand LineKick::moveToPreKick(
+  const World & world,
+  const Robot & robot)
+{
+  easy_move_to_.setPlannerOptions({});
+  easy_move_to_.setTargetPosition(getPreKickPosition(world));
+  easy_move_to_.face_travel();
+  std::vector<ateam_geometry::AnyShape> obstacles = {
+    ateam_geometry::makeCircle(world.ball.pos, 0.02)
+  };
+  return easy_move_to_.runFrame(robot, world, obstacles);
+}
+
+ateam_msgs::msg::RobotMotionCommand LineKick::faceBall(const World & world, const Robot & robot)
+{
+  easy_move_to_.setPlannerOptions({});
+  easy_move_to_.setTargetPosition(robot.pos);
+  easy_move_to_.face_point(target_point_);
+  easy_move_to_.setMaxAngularVelocity(1.0);
```

Comment:
Curious, do we ever change this back to the default? Does it change automatically?

Reply:
This affects MotionController btw

Reply:
In general we need to check whether we reset the motion planner options/settings or reinitialize those objects later